src/librustc_mir/transform/deaggregator.rs

```diff
@@ -7,6 +7,10 @@ pub struct Deaggregator;
 
 impl<'tcx> MirPass<'tcx> for Deaggregator {
     fn run_pass(&self, tcx: TyCtxt<'tcx>, _source: MirSource<'tcx>, body: &mut BodyAndCache<'tcx>) {
+        if tcx.sess.opts.debugging_opts.mir_opt_level == 0 {
```

Comment:
I don't really call this an optimization, but rather a desugaring of something only necessary during borrowck just like `FakeEdges`, `FalseUnwind`, `FakeRead` and `SetDiscriminant` on non-enums. It should be possible for cg_ssa to ignore all `Aggregate`'s except for arrays.

Reply:
@wesleywiser did you check that Miri still works without this pass? We might have inadvertently relied on it.

Reply:
Miri seems to do the same as cg_ssa: https://github.com/rust-lang/rust/blob/5e9ebf4da3522c479aea9c242cb3686f8d6afa78/src/librustc_mir/interpret/step.rs#L177-L199

Reply:
I didn't no. I'm fine with removing this change since this isn't really an optimization pass. I'll also add a comment to `rustc_mir::transform::run_optimization_passes()` saying that this isn't an optimization (it's nestled between all of the actual optimizations so it's easy to overlook).

Reply:
Well if it is a "desugaring" pass it should be grouped with them in `run_optimization_passes`.

> (it's nestled between all of the actual optimizations so it's easy to overlook)

Is there a way to apply the mir_opt_level centrally in `run_optimization_passes`? Having to do that for each optimization like here seems pretty fragile.

Reply:
@bjorn3 so, this would work correctly in Miri?

Given that codegen and Miri both don't ICE when this is still present, I am not entirely sure what you are saying (that this is just desugaring) is right. It doesn't seem to be treated like that (also re: placement in `run_optimization_passes`).

>  It should be possible for cg_ssa to ignore all Aggregate's except for arrays.

That seems odd, why would we lower structs but not arrays?

Reply:
Cc @eddyb (see above)

Reply:
> @bjorn3 so, this would work correctly in Miri (with the pass gone)?

Yes it should

> Given that codegen and Miri both don't ICE when this is still present, I am not entirely sure what you are saying (that this is just desugaring) is right. It doesn't seem to be treated like that (also re: placement in run_optimization_passes).

The codegen for `Rvalue::Aggregate` probably only existed because the deaggregate pass didn't exist back then. In cg_clif I only handle `AggregateKind::Array`.

> That seems odd, why would we lower structs but not arrays?

Because of #48193 (10% regression in artificial test case)

https://github.com/rust-lang/rust/blob/5e9ebf4da3522c479aea9c242cb3686f8d6afa78/src/librustc_mir/transform/deaggregator.rs#L17

Reply:
> Is there a way to apply the mir_opt_level centrally in `run_optimization_passes`? Having to do that for each optimization like here seems pretty fragile.

@RalfJung I took a stab at this in 349f193. Lmk what you think

Reply:
That seems better!

`run_optimization_passes` is still pretty confusing though. At least the name is bad, as not all passes run there are optimization passes. (I remember this confused the heck out of me when I first dug into this code.)

I imagined the way it would look is that there is first some list of passes that is always run, and then if mir-opt-level == 0 we return, else we continue with more passes. That way not each pass has to make that decision itself. (More fine-grained control, such as higher opt-levels or what const-prop is doing, still happens locally in the passes.) But maybe that would be too messy?

Reply:
We could do that but we'd need 3 lists of passes to retain the current ordering as-is (`RemoveNoopLandingPads` and `add_call_guards::CriticalCallEdges` are both required passes and happen right before codegen). Do you still like that better than this approach?

Reply:
I'm not sure. Let's see what others think.

Reply:
@oli-obk Do you have a preference here?

Reply:
I think having three lists is a good thing. It makes it very clear what's going on:

1. pre-optimization transformations
2. optimizations
3. post-optimization transformations

Reply:
Hmm, I wonder if 1 should be "post-borrowck" and 3 should be "pre-codegen".

As in, the only reason they exist *at that point* is that e.g. borrowck needs the "undesugared" form or that codegen needs/wants something that's not necessary/useful/good for everyone else.

Reply:
Ok, so how about with

1. post-borrowck-cleanup
2. optimizations
3. pre-codegen-cleanup

To make it clear that the pre/post optimizations are essentially just simplifying the MIR by removing things that are not necessary for the next phase?